<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Exception Handling &mdash; PyTango v7.1.0 documentation</title>
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '7.1.0',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="PyTango v7.1.0 documentation" href="index.html" />
    <link rel="next" title="FAQ" href="faq.html" />
    <link rel="prev" title="Util" href="server/util.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="faq.html" title="FAQ"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="server/util.html" title="Util"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">PyTango v7.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo.png" alt="Logo"/>
            </a></p>
            <h3><a href="index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="">Exception Handling</a><ul>
<li><a class="reference external" href="#exception-definition">Exception definition</a></li>
<li><a class="reference external" href="#throwing-exception-in-a-device-server">Throwing exception in a device server</a></li>
<li><a class="reference external" href="#exception-api">Exception API</a></li>
</ul>
</li>
</ul>

            <h4>Previous topic</h4>
            <p class="topless"><a href="server/util.html"
                                  title="previous chapter">Util</a></p>
            <h4>Next topic</h4>
            <p class="topless"><a href="faq.html"
                                  title="next chapter">FAQ</a></p>
            <h3>This Page</h3>
            <ul class="this-page-menu">
              <li><a href="_sources/exception.txt"
                     rel="nofollow">Show Source</a></li>
            </ul>
          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="exception-handling">
<h1>Exception Handling<a class="headerlink" href="#exception-handling" title="Permalink to this headline">¶</a></h1>
<div class="section" id="exception-definition">
<h2>Exception definition<a class="headerlink" href="#exception-definition" title="Permalink to this headline">¶</a></h2>
<p>All the exceptions that can be thrown by the underlying Tango C++ API are available
in the PyTango python module. Hence a user can catch one of the following
exceptions:</p>
<blockquote>
<ul class="simple">
<li><a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">DevFailed</span></tt></a></li>
<li><a title="PyTango.ConnectionFailed" class="reference internal" href="#PyTango.ConnectionFailed"><tt class="xref docutils literal"><span class="pre">ConnectionFailed</span></tt></a></li>
<li><a title="PyTango.CommunicationFailed" class="reference internal" href="#PyTango.CommunicationFailed"><tt class="xref docutils literal"><span class="pre">CommunicationFailed</span></tt></a></li>
<li><a title="PyTango.WrongNameSyntax" class="reference internal" href="#PyTango.WrongNameSyntax"><tt class="xref docutils literal"><span class="pre">WrongNameSyntax</span></tt></a></li>
<li><a title="PyTango.NonDbDevice" class="reference internal" href="#PyTango.NonDbDevice"><tt class="xref docutils literal"><span class="pre">NonDbDevice</span></tt></a></li>
<li><a title="PyTango.WrongData" class="reference internal" href="#PyTango.WrongData"><tt class="xref docutils literal"><span class="pre">WrongData</span></tt></a></li>
<li><a title="PyTango.NonSupportedFeature" class="reference internal" href="#PyTango.NonSupportedFeature"><tt class="xref docutils literal"><span class="pre">NonSupportedFeature</span></tt></a></li>
<li><a title="PyTango.AsynCall" class="reference internal" href="#PyTango.AsynCall"><tt class="xref docutils literal"><span class="pre">AsynCall</span></tt></a></li>
<li><a title="PyTango.AsynReplyNotArrived" class="reference internal" href="#PyTango.AsynReplyNotArrived"><tt class="xref docutils literal"><span class="pre">AsynReplyNotArrived</span></tt></a></li>
<li><a title="PyTango.EventSystemFailed" class="reference internal" href="#PyTango.EventSystemFailed"><tt class="xref docutils literal"><span class="pre">EventSystemFailed</span></tt></a></li>
<li><a title="PyTango.NamedDevFailedList" class="reference internal" href="#PyTango.NamedDevFailedList"><tt class="xref docutils literal"><span class="pre">NamedDevFailedList</span></tt></a></li>
<li><a title="PyTango.DeviceUnlocked" class="reference internal" href="#PyTango.DeviceUnlocked"><tt class="xref docutils literal"><span class="pre">DeviceUnlocked</span></tt></a></li>
</ul>
</blockquote>
<p>When an exception is caught, the sys.exc_info() function returns a tuple of three
values that give information about the exception that is currently being handled.
The values returned are (type, value, traceback).
Since most functions don&#8217;t need access to the traceback, the best solution is to
use something like exctype, value = sys.exc_info()[:2] to extract only the exception
type and value. If one of the Tango exceptions is caught, the exctype will be class
name of the exception (DevFailed, .. etc) and the value a tuple of dictionary objects
all of which containing the following kind of key-value pairs:</p>
<ul class="simple">
<li><strong>reason</strong>: a string describing the error type (more readable than the associated error code)</li>
<li><strong>desc</strong>: a string describing in plain text the reason of the error.</li>
<li><strong>origin</strong>: a string giving the name of the (C++ API) method which thrown the exception</li>
<li><strong>severity</strong>: one of the strings WARN, ERR, PANIC giving severity level of the error.</li>
</ul>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></td><td class="code"><div class="highlight"><pre><span class="c">#  Protect the script from Exceptions raised by the Tango or python itself</span>
<span class="k">try</span><span class="p">:</span>
    <span class="c"># Get proxy on the tangotest1 device</span>
    <span class="k">print</span> <span class="s">&quot;Getting DeviceProxy &quot;</span>
    <span class="n">tangotest</span> <span class="o">=</span> <span class="n">DeviceProxy</span><span class="p">(</span><span class="s">&quot;tango/tangotest/1&quot;</span><span class="p">)</span>

<span class="c">#Catch Tango and Systems  Exceptions</span>
<span class="k">except</span> <span class="n">DevFailed</span><span class="p">:</span>
    <span class="n">exctype</span> <span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">exc_info</span><span class="p">()[:</span><span class="mf">2</span><span class="p">]</span>
    <span class="k">print</span> <span class="s">&quot;Failed with exception ! &quot;</span> <span class="p">,</span> <span class="n">exctype</span>
    <span class="k">for</span> <span class="n">err</span> <span class="ow">in</span> <span class="n">value</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&quot; reason&quot;</span> <span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">reason</span>
        <span class="k">print</span> <span class="s">&quot; description&quot;</span> <span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">desc</span>
        <span class="k">print</span> <span class="s">&quot; origin&quot;</span> <span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">origin</span>
        <span class="k">print</span> <span class="s">&quot; severity&quot;</span> <span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">severity</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="throwing-exception-in-a-device-server">
<h2>Throwing exception in a device server<a class="headerlink" href="#throwing-exception-in-a-device-server" title="Permalink to this headline">¶</a></h2>
<p>The C++ Tango::Except class with its most important methods have been wrapped to Python.
Therefore, in a Python device server, you have the following methods to throw, re-throw or
print a Tango::DevFailed exception :</p>
<ul class="simple">
<li><em>throw_exception()</em> which is a static method</li>
<li><em>re_throw_exception()</em> which is also a static method</li>
<li><em>print_exception()</em> which is also a static method</li>
</ul>
<p>The following code is an example of a command method requesting a command on a sub-device and re-throwing
the exception in case of:</p>
<div class="highlight-python"><table class="highlighttable"><tr><td class="linenos"><pre>1
2
3
4
5
6
7</pre></td><td class="code"><div class="highlight"><pre><span class="k">try</span><span class="p">:</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">command_inout</span><span class="p">(</span><span class="s">&quot;SubDevCommand&quot;</span><span class="p">)</span>
<span class="k">except</span> <span class="n">PyTango</span><span class="o">.</span><span class="n">DevFailed</span><span class="p">,</span> <span class="n">e</span><span class="p">:</span>
    <span class="n">PyTango</span><span class="o">.</span><span class="n">Except</span><span class="o">.</span><span class="n">re_throw_exception</span><span class="p">(</span><span class="n">e</span><span class="p">,</span>
        <span class="s">&quot;MyClass_CommandFailed&quot;</span><span class="p">,</span>
        <span class="s">&quot;Sub device command SubdevCommand failed&quot;</span><span class="p">,</span>
        <span class="s">&quot;Command()&quot;</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">line 2:</th><td class="field-body">Send the command to the sub device in a try/catch block</td>
</tr>
<tr class="field"><th class="field-name">line 4-6:</th><td class="field-body">Re-throw the exception and add a new level of information in the exception stack</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="exception-api">
<h2>Exception API<a class="headerlink" href="#exception-api" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="PyTango.Except">
<em class="property">
class </em><tt class="descclassname">PyTango.</tt><tt class="descname">Except</tt><a class="headerlink" href="#PyTango.Except" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref docutils literal"><span class="pre">Boost.Python.instance</span></tt></p>
<dl class="docutils">
<dt>A containner for the static methods:</dt>
<dd><blockquote class="first">
<ul class="simple">
<li>throw_exception</li>
<li>re_throw_exception</li>
<li>print_exception</li>
<li>compare_exception</li>
</ul>
</blockquote>
<dl class="last staticmethod">
<dt id="PyTango.Except.print_exception">
<em class="property">
static </em><tt class="descname">print_exception</tt><big>(</big><em>exception</em><big>)</big><a class="headerlink" href="#PyTango.Except.print_exception" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints the Tango exception on the standard output</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameter:</th><td class="field-body"><em>exception</em> (<a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">DevFailed</span></tt></a>) &#8211; tango exception</td>
</tr>
<tr class="field"><th class="field-name">Return type:</th><td class="field-body">None</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="PyTango.DevError">
<em class="property">
class </em><tt class="descclassname">PyTango.</tt><tt class="descname">DevError</tt><a class="headerlink" href="#PyTango.DevError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref docutils literal"><span class="pre">Boost.Python.instance</span></tt></p>
<p>Structure describing any error resulting from a command execution,
or an attribute query, with following members:</p>
<blockquote>
<ul class="simple">
<li>reason : (<tt class="xref docutils literal"><span class="pre">str</span></tt>)</li>
<li>severity : (<a title="PyTango.ErrSeverity" class="reference external" href="client/other.html#PyTango.ErrSeverity"><tt class="xref docutils literal"><span class="pre">ErrSeverity</span></tt></a>) error severty (WARN, ERR, PANIC)</li>
<li>desc : (<tt class="xref docutils literal"><span class="pre">str</span></tt>) error description</li>
<li>origin : (<tt class="xref docutils literal"><span class="pre">str</span></tt>) Tango server method in which the error happened</li>
</ul>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.DevFailed">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">DevFailed</tt><a class="headerlink" href="#PyTango.DevFailed" title="Permalink to this definition">¶</a></dt>
<dd>Bases: <tt class="xref docutils literal"><span class="pre">exceptions.Exception</span></tt></dd></dl>

<dl class="exception">
<dt id="PyTango.ConnectionFailed">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">ConnectionFailed</tt><a class="headerlink" href="#PyTango.ConnectionFailed" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception is thrown when a problem occurs during the connection
establishment between the application and the device. The API is stateless.
This means that DeviceProxy constructors filter most of the exception
except for cases described in the following table.</p>
<p>The desc DevError structure field allows a user to get more precise information. These informations are :</p>
<dl class="docutils">
<dt><strong>DB_DeviceNotDefined</strong></dt>
<dd>The name of the device not defined in the database</dd>
<dt><strong>API_CommandFailed</strong></dt>
<dd>The device and command name</dd>
<dt><strong>API_CantConnectToDevice</strong></dt>
<dd>The device name</dd>
<dt><strong>API_CorbaException</strong></dt>
<dd>The name of the CORBA exception, its reason, its locality, its completed
flag and its minor code</dd>
<dt><strong>API_CantConnectToDatabase</strong></dt>
<dd>The database server host and its port number</dd>
<dt><strong>API_DeviceNotExported</strong></dt>
<dd>The device name</dd>
</dl>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.CommunicationFailed">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">CommunicationFailed</tt><a class="headerlink" href="#PyTango.CommunicationFailed" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception is thrown when a communication problem is detected during
the communication between the client application and the device server. It
is a two levels Tango::DevError structure. In case of time-out, the DevError
structures fields are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="23%" />
<col width="57%" />
<col width="12%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Level</th>
<th class="head">Reason</th>
<th class="head">Desc</th>
<th class="head">Severity</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>0</td>
<td>API_CorbaException</td>
<td>CORBA exception fields translated into a string</td>
<td>ERR</td>
</tr>
<tr><td>1</td>
<td>API_DeviceTimedOut</td>
<td>String with time-out value and device name</td>
<td>ERR</td>
</tr>
</tbody>
</table>
<p>For all other communication errors, the DevError structures fields are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="27%" />
<col width="55%" />
<col width="11%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Level</th>
<th class="head">Reason</th>
<th class="head">Desc</th>
<th class="head">Severity</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>0</td>
<td>API_CorbaException</td>
<td>CORBA exception fields translated into a string</td>
<td>ERR</td>
</tr>
<tr><td>1</td>
<td>API_CommunicationFailed</td>
<td>String with device, method, command/attribute name</td>
<td>ERR</td>
</tr>
</tbody>
</table>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.WrongNameSyntax">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">WrongNameSyntax</tt><a class="headerlink" href="#PyTango.WrongNameSyntax" title="Permalink to this definition">¶</a></dt>
<dd>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></dd></dl>

<p>This exception has only one level of Tango::DevError structure. The possible
value for the reason field are :</p>
<blockquote>
<dl class="docutils">
<dt><strong>API_UnsupportedProtocol</strong></dt>
<dd>This error occurs when trying to build a DeviceProxy or an AttributeProxy
instance for a device with an unsupported protocol. Refer to the appendix
on device naming syntax to get the list of supported database modifier</dd>
<dt><strong>API_UnsupportedDBaseModifier</strong></dt>
<dd>This error occurs when trying to build a DeviceProxy or an AttributeProxy
instance for a device/attribute with a database modifier unsupported.
Refer to the appendix on device naming syntax to get the list of
supported database modifier</dd>
<dt><strong>API_WrongDeviceNameSyntax</strong></dt>
<dd>This error occurs for all the other error in device name syntax. It is
thrown by the DeviceProxy class constructor.</dd>
<dt><strong>API_WrongAttributeNameSyntax</strong></dt>
<dd>This error occurs for all the other error in attribute name syntax. It
is thrown by the AttributeProxy class constructor.</dd>
<dt><strong>API_WrongWildcardUsage</strong></dt>
<dd>This error occurs if there is a bad usage of the wildcard character</dd>
</dl>
</blockquote>
<dl class="exception">
<dt id="PyTango.NonDbDevice">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">NonDbDevice</tt><a class="headerlink" href="#PyTango.NonDbDevice" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
This exception has only one level of Tango::DevError structure. The reason
field is set to API_NonDatabaseDevice. This exception is thrown by the API
when using the DeviceProxy or AttributeProxy class database access for
non-database device.</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.WrongData">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">WrongData</tt><a class="headerlink" href="#PyTango.WrongData" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception has only one level of Tango::DevError structure.
The possible value for the reason field are :</p>
<dl class="docutils">
<dt><strong>API_EmptyDbDatum</strong></dt>
<dd>This error occurs when trying to extract data from an empty DbDatum
object</dd>
<dt><strong>API_IncompatibleArgumentType</strong></dt>
<dd>This error occurs when trying to extract data with a type different
than the type used to send the data</dd>
<dt><strong>API_EmptyDeviceAttribute</strong></dt>
<dd>This error occurs when trying to extract data from an empty
DeviceAttribute object</dd>
<dt><strong>API_IncompatibleAttrArgumentType</strong></dt>
<dd>This error occurs when trying to extract attribute data with a type
different than the type used to send the data</dd>
<dt><strong>API_EmptyDeviceData</strong></dt>
<dd>This error occurs when trying to extract data from an empty DeviceData
object</dd>
<dt><strong>API_IncompatibleCmdArgumentType</strong></dt>
<dd>This error occurs when trying to extract command data with a type
different than the type used to send the data</dd>
</dl>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.NonSupportedFeature">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">NonSupportedFeature</tt><a class="headerlink" href="#PyTango.NonSupportedFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
This exception is thrown by the API layer when a request to a feature
implemented in Tango device interface release n is requested for a device
implementing Tango device interface n-x. There is one possible value for
the reason field which is API_UnsupportedFeature.</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.AsynCall">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">AsynCall</tt><a class="headerlink" href="#PyTango.AsynCall" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception is thrown by the API layer when a the asynchronous model id
badly used. This exception has only one level of Tango::DevError structure.
The possible value for the reason field are :</p>
<dl class="docutils">
<dt><strong>API_BadAsynPollId</strong></dt>
<dd>This error occurs when using an asynchronous request identifier which is not
valid any more.</dd>
<dt><strong>API_BadAsyn</strong></dt>
<dd>This error occurs when trying to fire callback when no callback has been
previously registered</dd>
<dt><strong>API_BadAsynReqType</strong></dt>
<dd>This error occurs when trying to get result of an asynchronous request with
an asynchronous request identifier returned by a non-coherent asynchronous
request (For instance, using the asynchronous request identifier returned
by a command_inout_asynch() method with a read_attribute_reply() attribute).</dd>
</dl>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.AsynReplyNotArrived">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">AsynReplyNotArrived</tt><a class="headerlink" href="#PyTango.AsynReplyNotArrived" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception is thrown by the API layer when:</p>
<blockquote>
<ul class="simple">
<li>a request to get asynchronous reply is made and the reply is not yet arrived</li>
<li>a blocking wait with timeout for asynchronous reply is made and the timeout expired.</li>
</ul>
</blockquote>
<p>There is one possible value for the reason field which is API_AsynReplyNotArrived.</p>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.EventSystemFailed">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">EventSystemFailed</tt><a class="headerlink" href="#PyTango.EventSystemFailed" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception is thrown by the API layer when subscribing or unsubscribing
from an event failed. This exception has only one level of Tango::DevError
structure. The possible value for the reason field are :</p>
<dl class="docutils">
<dt><strong>API_NotificationServiceFailed</strong></dt>
<dd>This error occurs when the subscribe_event() method failed trying to
access the CORBA notification service</dd>
<dt><strong>API_EventNotFound</strong></dt>
<dd>This error occurs when you are using an incorrect event_id in the
unsubscribe_event() method</dd>
<dt><strong>API_InvalidArgs</strong></dt>
<dd>This error occurs when NULL pointers are passed to the subscribe or
unsubscribe event methods</dd>
<dt><strong>API_MethodArgument</strong></dt>
<dd>This error occurs when trying to subscribe to an event which has already
been subsribed to</dd>
<dt><strong>API_DSFailedRegisteringEvent</strong></dt>
<dd>This error means that the device server to which the device belongs to
failed when it tries to register the event. Most likely, it means that
there is no event property defined</dd>
<dt><strong>API_EventNotFound</strong></dt>
<dd>Occurs when using a wrong event identifier in the unsubscribe_event
method</dd>
</dl>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.DeviceUnlocked">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">DeviceUnlocked</tt><a class="headerlink" href="#PyTango.DeviceUnlocked" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></p>
<blockquote>
<p>This exception is thrown by the API layer when a device locked by the
process has been unlocked by an admin client. This exception has two levels
of Tango::DevError structure. There is only possible value for the reason
field which is</p>
<dl class="docutils">
<dt><strong>API_DeviceUnlocked</strong></dt>
<dd>The device has been unlocked by another client (administration client)</dd>
</dl>
<p>The first level is the message reported by the Tango kernel from the server
side. The second layer is added by the client API layer with informations on
which API call generates the exception and device name.</p>
</blockquote>
</dd></dl>

<dl class="exception">
<dt id="PyTango.NotAllowed">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">NotAllowed</tt><a class="headerlink" href="#PyTango.NotAllowed" title="Permalink to this definition">¶</a></dt>
<dd>Bases: <a title="PyTango.DevFailed" class="reference internal" href="#PyTango.DevFailed"><tt class="xref docutils literal"><span class="pre">PyTango.DevFailed</span></tt></a></dd></dl>

<dl class="exception">
<dt id="PyTango.NamedDevFailedList">
<em class="property">
exception </em><tt class="descclassname">PyTango.</tt><tt class="descname">NamedDevFailedList</tt><a class="headerlink" href="#PyTango.NamedDevFailedList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref docutils literal"><span class="pre">Boost.Python.instance</span></tt></p>
<blockquote>
<p>This exception is only thrown by the DeviceProxy::write_attributes()
method. In this case, it is necessary to have a new class of exception
to transfer the error stack for several attribute(s) which failed during
the writing. Therefore, this exception class contains for each attributes
which failed :</p>
<blockquote>
<ul class="simple">
<li>The name of the attribute</li>
<li>Its index in the vector passed as argumen tof the write_attributes() method</li>
<li>The error stack</li>
</ul>
</blockquote>
</blockquote>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="faq.html" title="FAQ"
             >next</a> |</li>
        <li class="right" >
          <a href="server/util.html" title="Util"
             >previous</a> |</li>
        <li><a href="index.html">PyTango v7.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
      &copy; Copyright 2009, ALBA - CELLS.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.2.
    </div>
  </body>
</html>